### PR TITLE
Derive randomness from a single seed

### PR DESCRIPTION
**Pull request by stechu**
For each coin (UTXO), now all the randomness is derived from single secret.

This would be useful for Manta's wallet protocols.

(This is a breaking change)

### CODE REVIEW

**Comment by ghzlatarev** [comment]
The integration check will fail, until https://github.com/Manta-Network/manta-types/pull/7 is merged.
The version of manta-crypto it pulls is different than the one in manta-crypto's manta branch.

**Comment by stechu** [comment]
@ghzlatarev  the integration will fall even if #7 is merged. This PR breaks `manta-api` as well.

**Comment by ghzlatarev** [comment]
> @ghzlatarev the integration will fail even if #7 is merged. This PR breaks `manta-api` as well.

Yeah you are right, didn't read your description..

File: manta-asset/src/sampling.rs [hunk 1]
```diff
@@ -42,12 +43,11 @@ pub trait Sampling {
 	/// - asset_id: self-explained
 	/// - value: the number of assets held in this token
 	/// - rng: self-explained
-	fn sample<R: RngCore + CryptoRng>(
+	fn sample<R: RngCore + CryptoRng + SeedableRng<Seed = Self::SecretKey>>(
```
Comment (Kevingislason) [inline manta-asset/src/sampling.rs:46]: I don’t think we need this generic R anymore. See main comment.

File: manta-asset/src/sampling.rs [hunk 2]
```diff
@@ -61,17 +61,17 @@ impl Sampling for MantaAsset {
...
 		param: &Self::Param,
 		secret_key: &Self::SecretKey,
 		asset_id: &AssetId,
 		value: &Self::Value,
-		rng: &mut R,
 	) -> Result<Self, MantaError> {
+		// In mantaPay, rng should be instantiated to ChaCha20Rng
+		let mut rng = R::from_seed(secret_key.clone());
```
Comment (Kevingislason) [inline manta-asset/src/sampling.rs:74]: I think we should strictly enforce this comment here in manta types, i.e. we should directly instantiate the ChaCha20Rng here. If an external user uses this library, and uses a different RNG besides ChaCha20, then they will get different addresses at the same HD key path. There should be just one canonical way to derive addresses.

Comment (stechu) [reply manta-asset/src/sampling.rs:74]: Right. This was designated for the future upgrade the protocol. I guess this should not be a worry in the anytime soon.

**Comment by stechu** [comment]
@Kevingislason  your comments have been addressed.

**Comment by stechu** [comment]
This closes #8.

**Review by Kevingislason — APPROVED** [review]

**Pull request merged**